# CGS4144
Bioinformatics Project

Possible sources:

https://www.refine.bio/experiments/SRP076307/single-cell-rna-seq-of-human-pancreatic-endocrine-cells-from-juvenile-adult-control-and-type-2-diabetic-donors#samples

backup:

https://www.refine.bio/experiments/SRP033566/comparative-rna-sequencing-analysis-of-myocardial-and-circulating-small-rnas-in-human-heart-failure-and-their-utility-as-biomarkers-small-rna-seq

# Some Questions:

Differential Expression: What genes are differentially expressed across donor types (child, adult control, type 1 diabetes, type 2 diabetes), and do they associate with specific biological pathways?

Gene Variability by Donor Type: How does gene expression variability differ between the different donor types?

Age-Related Gene Signatures: Is age a significant factor in shaping the gene expression profile in the dataset?

Diabetes-Specific Signatures: Are there unique gene expression signatures that differentiate type 1 and type 2 diabetes donors from controls?

Dimensionality Reduction: Can principal component analysis distinguish between the different donor types based on their gene expression profiles?

Machine Learning Classification: Is it possible to use machine learning algorithms to accurately classify donor types based on gene expression?

Gene Interaction Networks: Can a gene interaction network identify key genes central to the differences between donor types?

Metabolic and Signaling Pathways: Are there metabolic or signaling pathways disproportionately affected in diabetes donors?

Type 2 and Juvenile Similarity: What specific genes or pathways are similar in gene expression between juvenile cells and cells from type 2 diabetes donors?

Functional Validation: What in vitro or in vivo experiments could validate the functional roles of identified key genes?
