---
title: "differential expression analysis"
author: "Manuel Cortes"
date: "2023-09-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Human Annotation
library(org.Hs.eg.db)


library(tidyr)

set.seed(12345)
```

# Loading in the data

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP076307")

# Declare the file path to the gene expression matrix file
data_file <- file.path(data_dir, "SRP076307.tsv")

# Declare the file path to the metadata file
metadata_file <- file.path(data_dir, "metadata_SRP076307.tsv")

# Read in metadata TSV file
metadata <- readr::read_tsv(metadata_file)

# Read in data TSV file
expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")
```

# Preliminary metadata preprocessing

```{r}
# Make the data in the order of the metadata
expression_df <- expression_df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(expression_df), metadata$refinebio_accession_code)


# This will add a new column that contains the group to which the samples belong to
metadata <- metadata %>%
  dplyr::mutate(
    Group = dplyr::case_when(
      stringr::str_detect(refinebio_title, "AAJF122|ABAF490|ACAP236") ~ "Control",
      stringr::str_detect(refinebio_title, "ACGI428") ~ "Type1",
      stringr::str_detect(refinebio_title, "HP-15041|HP-15085") ~ "Type2",
      stringr::str_detect(refinebio_title, "ICRH76|ICRH80") ~ "Child",
      TRUE ~ "Other" # Just in case there are titles that do not match any of the above
    )
  )

# Extract unique values of the Group column, but exclude "Control" and "Other"
unique_groups <- setdiff(unique(metadata$Group), c("Control", "Other"))

# Make sure "Control" is the first level, followed by the other unique values, and "Other" is the last level
ordered_levels <- c("Control", sort(unique_groups), "Other")

# Convert the Group column to a factor with these levels
metadata <- metadata %>%
  dplyr::mutate(Group = factor(Group, levels = ordered_levels))

# dplyr::select(metadata, refinebio_title, Group)

metadata %>%
  dplyr::count(Group) %>%
  tidyr::as_tibble()


```

# Preliminary data preprocessing

```{r}
# Define a minimum counts cutoff and filter the data to include
# only rows (genes) that have total counts above the cutoff
filtered_expression_df <- expression_df %>%
  dplyr::filter(rowSums(.) >= 10)

gene_matrix <- round(filtered_expression_df)

ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = gene_matrix,
  # Supply the `colData` with our metadata data frame
  colData = metadata,
  # Supply our experimental variable to `design`
  design = ~Group
)
```

# Running DESeq

```{r}
deseq_object <- DESeq(ddset)
```
```{r}
resultsNames(deseq_object)
```

# Analysis

anything beyond this is not very stable

```{r}
deseq_results <- results(deseq_object, contrast = c("Group", "Type2", "Control"))

deseq_results <- lfcShrink(
  deseq_object,
  coef = "Group_Type2_vs_Control", # Using the correct coefficient name
  res = deseq_results
)

head(deseq_results)


# this is of class DESeqResults -- we want a data frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are row names -- let's make them a column for easy display
  tibble::rownames_to_column("Gene") %>%
  # add a column for significance threshold results
  dplyr::mutate(threshold = padj < 0.05) %>%
  # sort by statistic -- the highest values will be genes with
  # higher expression in RPL10 mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))

head(deseq_df)

readr::write_tsv(
  deseq_df,
  file.path(
    "results",
    "SRP076307_diff_expr_results_ensembl.tsv" # Replace with a relevant output file name
  )
)
```
