---
title: "differential expression analysis"
author: "Manuel Cortes"
date: "2023-09-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Human Annotation
library(org.Hs.eg.db)

set.seed(12345)
```

# Loading in the data

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP033566")

# Declare the file path to the gene expression matrix file
data_file <- file.path(data_dir, "SRP033566.tsv")

# Declare the file path to the metadata file
metadata_file <- file.path(data_dir, "metadata_SRP033566.tsv")

# Read in metadata TSV file
metadata <- readr::read_tsv(metadata_file)

# Read in data TSV file
expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")
```

# Preliminary metadata preprocessing

```{r}
# Make the data in the order of the metadata
expression_df <- expression_df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(expression_df), metadata$refinebio_accession_code)

#This will add a new column that contains the group to which the samples belong to
metadata <- metadata %>%
  dplyr::mutate(Group = stringr::str_extract(refinebio_title, "^[A-Z]{3}"))

# Extract unique values of the Group column, but exclude "CTL"
unique_groups <- setdiff(unique(metadata$Group), "CTL")

# Make sure "CTL" is the first level, followed by the other unique values
ordered_levels <- c("CTL", sort(unique_groups))

# Convert the Group column to a factor with these levels
metadata <- metadata %>%
  dplyr::mutate(Group = factor(Group, levels = ordered_levels))

dplyr::select(metadata, refinebio_title, Group)


```

# Preliminary data preprocessing

```{r}
# Define a minimum counts cutoff and filter the data to include
# only rows (genes) that have total counts above the cutoff
filtered_expression_df <- expression_df %>%
  dplyr::filter(rowSums(.) >= 10)

gene_matrix <- round(filtered_expression_df)

ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = gene_matrix,
  # Supply the `colData` with our metadata data frame
  colData = metadata,
  # Supply our experimental variable to `design`
  design = ~Group
)
```

# Running DESeq

```{r}
deseq_object <- DESeq(ddset)
```

# Analysis

anything beyond this is not very stable

```{r}
deseq_results <- results(deseq_object, contrast = c("Group", "DCM", "CTL"))

deseq_results <- lfcShrink(
  deseq_object, # The original DESeq2 object after running DESeq()
  coef = 2, # The log fold change coefficient used in DESeq(); the default is 2.
  res = deseq_results # The original DESeq2 results table
)

head(deseq_results)

# this is of class DESeqResults -- we want a data frame
deseq_df <- deseq_results %>%
  # make into data.frame
  as.data.frame() %>%
  # the gene names are row names -- let's make them a column for easy display
  tibble::rownames_to_column("Gene") %>%
  # add a column for significance threshold results
  dplyr::mutate(threshold = padj < 0.05) %>%
  # sort by statistic -- the highest values will be genes with
  # higher expression in RPL10 mutated samples
  dplyr::arrange(dplyr::desc(log2FoldChange))

head(deseq_df)

readr::write_tsv(
  deseq_df,
  file.path(
    "results",
    "SRP033566_diff_expr_results.tsv" # Replace with a relevant output file name
  )
)

```


# Enrichment analysis

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("GenomicSuperSignature")

```


```{r}

library(clusterProfiler)
library(biomaRt)  # For converting Ensembl IDs to Entrez IDs
library(clusterProfiler)
library(GenomicSuperSignature)

# Filter for significantly differentially expressed genes
sig_genes_df <- deseq_df %>%
  dplyr::filter(padj < 0.05)  # Adjust threshold as needed

# # Optionally, you can also filter by log2 fold change, for example, |log2FC| > 1
# sig_genes_df <- sig_genes_df %>%
#   dplyr::filter(abs(log2FoldChange) > 1)  # Adjust threshold as needed

# # Filter for significantly differentially expressed genes using the threshold column
# sig_genes_df <- deseq_df %>%
#   dplyr::filter(threshold == TRUE)

# Extract the list of significant gene names/IDs
ensembl_ids <- sig_genes_df$Gene

ensembl_ids <- unique(rownames(deseq_object))



# Map Ensembl IDs to their associated Entrez IDs
entrez_ids <- mapIds(
  org.Hs.eg.db, # Replace with annotation package for your organism
  keys = ensembl_ids,
  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
  column = "ENTREZID", # The type of gene identifiers you would like to map to
  multiVals = "list"
)
# Check for any NA or missing values and remove them
# entrez_ids <- na.omit(entrez_ids)

```
```{r}

kegg_enrich <- enrichKEGG(gene = entrez_ids,
                          organism = 'hsa',  # Human genes
                          keyType = 'kegg',
                          pAdjustMethod = 'BH',  # Benjamini-Hochberg adjustment
                          qvalueCutoff = 0.05)  # Adjust as needed
dotplot(kegg_enrich, showCategory = 10) + theme_minimal()

library(DOSE)
do_res <- enrichDO(gene = entrez_ids, pAdjustMethod = "BH", pvalueCutoff = 0.05)
dotplot(do_res) + theme_minimal()

```
