---
title: "differential expression analysis"
author: "Manuel Cortes"
date: "2023-09-21"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# Attach the DESeq2 library
library(DESeq2)

# Attach the ggplot2 library for plotting
library(ggplot2)

# We will need this so we can use the pipe: %>%
library(magrittr)

# Human Annotation
library(org.Hs.eg.db)

set.seed(12345)
```

# Loading in the data

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP076307")

# Declare the file path to the gene expression matrix file
data_file <- file.path(data_dir, "SRP076307.tsv")

# Declare the file path to the metadata file
metadata_file <- file.path(data_dir, "metadata_SRP076307.tsv")

# Read in metadata TSV file
metadata <- readr::read_tsv(metadata_file)

# Read in data TSV file
expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")
```

# Preliminary metadata preprocessing

```{r}
# Make the data in the order of the metadata
expression_df <- expression_df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(expression_df), metadata$refinebio_accession_code)

#This will add a new column that contains the group to which the samples belong to
metadata <- metadata %>%
  dplyr::mutate(Group = stringr::str_extract(refinebio_title, "^[A-Z]{3}"))

# Extract unique values of the Group column, but exclude "CTL"
unique_groups <- setdiff(unique(metadata$Group), "CTL")

# Make sure "CTL" is the first level, followed by the other unique values
ordered_levels <- c("CTL", sort(unique_groups))

# Convert the Group column to a factor with these levels
metadata <- metadata %>%
  dplyr::mutate(Group = factor(Group, levels = ordered_levels))

dplyr::select(metadata, refinebio_title, Group)


```

# Preliminary data preprocessing

```{r}
# Define a minimum counts cutoff and filter the data to include
# only rows (genes) that have total counts above the cutoff
filtered_expression_df <- expression_df %>%
  dplyr::filter(rowSums(.) >= 10)

gene_matrix <- round(filtered_expression_df)

ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = gene_matrix,
  # Supply the `colData` with our metadata data frame
  colData = metadata,
  # Supply our experimental variable to `design`
  design = ~Group
)
```

# Analysis

# For DESeq, please see differential-expression-analysis.Rmd.
# Here, we will simply load the result from it

```{r}

# Define the file path
file_path <- file.path("results", "SRP033566_diff_expr_results.tsv")

# Read the file
deseq_df <- readr::read_tsv(file_path)

```

# Enrichment analysis

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("clusterProfiler")
BiocManager::install("biomaRt")

```


```{r}

library(biomaRt)  # For converting Ensembl IDs to Entrez IDs
library(clusterProfiler)



# # Filter for significantly differentially expressed genes using the threshold column
# sig_genes_df <- deseq_df %>%
#   dplyr::filter(threshold == TRUE)
# entrez_ids <- sig_genes_df$Gene


# Extract the list of significant gene names/IDs
ensembl_ids <- deseq_df$Gene

# Map Ensembl IDs to their associated Entrez IDs
entrez_ids <- mapIds(
  org.Hs.eg.db, # Replace with annotation package for your organism
  keys = ensembl_ids,
  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
  column = "ENTREZID", # The type of gene identifiers you would like to map to
  multiVals = "list"
)
# Check for any NA or missing values and remove them
entrez_ids <- na.omit(entrez_ids)
```



```{r}

kegg_enrich <- enrichKEGG(gene = entrez_ids,
                          organism = 'hsa',  # Human genes
                          keyType = 'kegg',
                          pAdjustMethod = 'BH',  # Benjamini-Hochberg adjustment
                          qvalueCutoff = 0.05)  # Adjust as needed
kegg_enrich_df <- as.data.frame(kegg_enrich)
head(kegg_enrich)
write.csv(kegg_enrich_df, "results/kegg_enrichment_results.csv", row.names = FALSE)


dotplot(kegg_enrich, showCategory = 10) + theme_minimal()
```


```{r}
library(DOSE)
do_res <- enrichDO(gene = entrez_ids, pAdjustMethod = "BH", pvalueCutoff = 0.05)

# Convert the enrichment object to a data frame
do_res_df <- as.data.frame(do_res)
# View the top rows of the table
head(do_res_df)
write.csv(do_res_df, "results/do_enrichment_results.csv", row.names = FALSE)


dotplot(do_res) + theme_minimal()

```
